### PR TITLE
Expose ConcurrentDictionary GetOrAdd/AddOrUpdate overloads in netcoreapp1.1

### PR DESCRIPTION
**Pull request by justinvp**
Port #1783 from future to master. Contributes to #2869.

I cherry-picked (and slightly tweaked) the two relevant commits from the future branch:
 - fffd49fe1c35544b388322608d1e16374f3832e6 (fixed the conflict due to moved test file)
 - ee5906f35222b80a729541683f76f43fe2c49204 (updated to use current `throw null;` style)

I then added a third commit that cleaned some things up, extracted tests, and exposed the APIs in netcoreapp1.1.

cc: @stephentoub

### CODE REVIEW

**Comment by justinvp** [comment]
@ericstj, is there anything you think I should change in this PR to bring it in-line with what you did in #13199?

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj [hunk 1]
```diff
@@ -5,6 +5,7 @@
     <AssemblyVersion>4.0.10.0</AssemblyVersion>
```
Comment (JonHanna) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: Should the version not go up if public members are added?

File: src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryTests.cs [hunk 2]
```diff
@@ -441,25 +444,47 @@ private static void TestGetOrAddOrUpdate(int cLevel, int initSize, int threads,
...
+                                        case 0:
+                                            dict.GetOrAdd(j, -j);
+                                            break;
+                                        case 1:
+                                            dict.GetOrAdd(j, x => -x);
+                                            break;
+                                        case 2:
+                                            if (s_getOrAddNetCoreApp11 != null)
+                                            {
+                                                s_getOrAddNetCoreApp11(dict, j);
```
Comment (stephentoub) [inline src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryTests.cs]: An alternative would be to implement the new method behavior as extension methods in the test that are compiled in only when testing the older version, and then leave this test itself clean.

Comment (justinvp) [inline src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryTests.cs]: Done.

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj (hunk 1 above)
Comment (justinvp) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: I bumped the version (though, I'm not confident I changed it in all the right places). Tests pass locally, but CI is now failing across the board (previously was green) and I'm not sure what the issue is. Hmm.

Comment (JonHanna) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: I was as much asking because I'm not confident I bumped the right places on something I'm working on :(

**Comment by justinvp** [comment]
Since this is adding new APIs, the minor version should be incremented.

What files need to be updated to bump the version to 4.2.0.0?

I tried updating the following:

 - [src/System.Collections.Concurrent/dir.props](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/dir.props)
  - Changing `<AssemblyVersion>4.0.14.0</AssemblyVersion>` => `4.2.0.0`
 - [src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj)
  - Changing `<AssemblyVersion>4.0.10.0</AssemblyVersion>` => `4.2.0.0`
 - [src/System.Collections.Concurrent/src/System.Collections.Concurrent.csproj](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System.Collections.Concurrent.csproj)
  - Adding `<AssemblyVersion>4.2.0.0</AssemblyVersion>`

Tests passed for me locally, but CI was failing across the board and I couldn't find a reason for the failures. So I've reverted the version changes for now (CI is back to green).

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj (hunk 1 above)
Comment (JonHanna) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: Well since the last change I tried there was copying something here, that's like Pooh and Piglet hunting Woozles and following their own footprints around a tree.

**Comment by stephentoub** [comment]
Thanks, Justin.  The src/test changes LGTM.  @weshaggard or @joperezr could probably help with the ref/version stuff.

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj (hunk 1 above)
Comment (joperezr) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: You do have to upgrade the version at least for the netcoreapp1.1 framework since you are adding API there. To do so, please update src/System.Collections.Concurrent/dir.props to have it's assembly version be 4.1.0.0. Also, you will need to keep this declaration in here, but you will need to add a Condition like:
```xml
<AssemblyVersion Condition="'$(TargetGroup)'=='netstandard1.3'">4.0.10.0</AssemblyVersion>
```

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.builds [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <Project Include="System.Collections.Concurrent.csproj" />
+    <Project Include="System.Collections.Concurrent.csproj">
+      <TargetGroup>netcoreapp1.1</TargetGroup>
```
Comment (joperezr) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.builds]: Could you please make netcoreapp1.1 the default? that wold mean changing the .csproj to have netcoreapp1.1 as the nuget target moniker by default, and in here you will be cross compiling for netstandard1.3 instead. The reason to do this is because you want the default build of the .csproj to be the newest version of it.

File: src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds [hunk 4]
```diff
@@ -8,6 +8,10 @@
       <TestTFMs>netcoreapp1.0</TestTFMs>
     </Project>
     <Project Include="System.Collections.Concurrent.Tests.csproj">
+      <TargetGroup>netcoreapp1.1</TargetGroup>
+      <TestTFMs>netcoreapp1.1</TestTFMs>
```
Comment (joperezr) [inline src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds]: > <TestTFMs>netcoreapp1.1</TestTFMs> [](start = 5, length = 35)

This is already the default so you don't need it.

**Comment by joperezr** [comment]
Looks like you are not crosscompiling the src for netcoreapp1.1 which is something that you have to fix. To do this, add a new configuration in your src/System.Collections.Concurrent.builds file for netcoreapp1.1, and then make sure that you are defining the `<ContractProject>` property for netstandard1.3, since you will need to pass in the targetGroup for that case.

**Comment by danmosemsft** [comment]
@justinvp have you had a chance to address feedback?

**Comment by danmosemsft** [comment]
oops

**Comment by justinvp** [comment]
@danmosemsft, sorry, I haven't had time to work through getting the version incremented. Will try to get to it in the next week or so.

**Comment by justinvp** [comment]
@joperezr, I made changes based on your feedback, but something still isn't right and I'm not familiar enough with the build/packaging to know what needs to be changed to address it.

Here's one of the errors at the end of msbuild.log from one of the CI builds:

> D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should support API version 4.1.0.0 on .NETCoreApp,Version=v1.0/win7-x86 but D:\j\workspace\windows_nt_de---4526f5ff\packages\System.Collections.Concurrent\4.3.0\ref\netstandard1.3\System.Collections.Concurrent.dll was found to support 4.0.10.0. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]

Any ideas?

File: src/System.Collections.Concurrent/pkg/System.Collections.Concurrent.pkgproj [hunk 5]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
-    <ProjectReference Include="..\ref\System.Collections.Concurrent.csproj">
+    <ProjectReference Include="..\ref\System.Collections.Concurrent.builds">
       <SupportedFramework>net46;netcore50;netcoreapp1.0;$(AllXamarinFrameworks)</SupportedFramework>
```
Comment (weshaggard) [inline src/System.Collections.Concurrent/pkg/System.Collections.Concurrent.pkgproj]: This should be updated to `<SupportedFramework>netcoreapp1.1;net463;$(AllXamarinFrameworks)</SupportedFramework>`

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.builds [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <Project Include="System.Collections.Concurrent.csproj" />
+    <Project Include="System.Collections.Concurrent.csproj">
+      <TargetGroup>netstandard1.3</TargetGroup>
```
Comment (weshaggard) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.builds]: Not 100% sure but you many need to remove this netstandard1.3 build configuration as well. It will automatically be harvested from the last stable package.

File: src/System.Collections.Concurrent/src/System.Collections.Concurrent.csproj [hunk 7]
```diff
@@ -14,6 +14,11 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|AnyCPU'" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net46_Debug|AnyCPU'" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net46_Release|AnyCPU'" />
+  <ItemGroup Condition="'$(TargetGroup)' == 'netstandard1.3'">
```
Comment (weshaggard) [inline src/System.Collections.Concurrent/src/System.Collections.Concurrent.csproj]: This isn't necessary as this project doesn't build a netstandard1.3 build configuration.

**Comment by justinvp** [comment]
Thanks, Wes. I made changes per your suggestions, but now seeing new errors in msbuild.log:

```
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3 but has no compile assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3 but has no runtime assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win-x86 but has no compile assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win-x86 but has no runtime assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win-x64 but has no compile assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win-x64 but has no runtime assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win7-x86 but has no compile assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win7-x86 but has no runtime assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win7-x64 but has no compile assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should be supported on .NETFramework,Version=v4.6.3/win7-x64 but has no runtime assets. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent should support API version 4.1.0.0 on .NETStandard,Version=v1.7 but D:\j\workspace\windows_nt_de---4526f5ff\packages\System.Collections.Concurrent\4.3.0\ref\netstandard1.3\System.Collections.Concurrent.dll was found to support 4.0.10.0. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
         D:\j\workspace\windows_nt_de---4526f5ff\Tools\Packaging.targets(1096,5): error : System.Collections.Concurrent has mismatched compile (4.0.10.0) and runtime (4.1.0.0) versions on .NETStandard,Version=v1.7. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Collections.Concurrent\pkg\System.Collections.Concurrent.pkgproj]
```

Any ideas?

File: src/System.Collections.Concurrent/ref/project.json [hunk 8]
```diff
@@ -4,6 +4,7 @@
     "System.Threading.Tasks": "4.0.11"
   },
   "frameworks": {
-    "netstandard1.3": {}
+    "netstandard1.3": {},
```
Comment (ianhays) [inline src/System.Collections.Concurrent/ref/project.json]: should this be netstandard1.7?

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj [hunk 9]
```diff
@@ -2,9 +2,10 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.10.0</AssemblyVersion>
+    <AssemblyVersion Condition="'$(TargetGroup)'=='netstandard1.3'">4.0.10.0</AssemblyVersion>
     <OutputType>Library</OutputType>
-    <NuGetTargetMoniker>.NETStandard,Version=v1.3</NuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(TargetGroup)'==''">.NETCoreApp,Version=v1.1</NuGetTargetMoniker>
+    <DefineConstants Condition="'$(TargetGroup)'==''">$(DefineConstants);netcoreapp11</DefineConstants>
```
Comment (ianhays) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: Are these ever false? The only entry in the builds file has a null TargetGroup.

File: src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Collections.Concurrent.Tests
+{
+    // Allows the ConcurrentDictionary tests to run on targets that do not have the new GetOrAdd/AddOrUpdate overloads.
```
Comment (ianhays) [inline src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryExtensions.cs]: Why? What's the purpose in testing a function that doesn't actually exist in that version?

File: src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds [hunk 11]
```diff
@@ -8,6 +8,9 @@
       <TestTFMs>netcoreapp1.0</TestTFMs>
     </Project>
     <Project Include="System.Collections.Concurrent.Tests.csproj">
+      <TargetGroup>netcoreapp1.1</TargetGroup>
```
Comment (ianhays) [inline src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds]: You may need to set TestTFMs.

File: src/System.Collections.Concurrent/dir.props [hunk 12]
```diff
@@ -2,6 +2,6 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="..\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.14.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
```
Comment (joperezr) [inline src/System.Collections.Concurrent/dir.props]: I talked with @ericstj and it looks like we intentionally keep AssemblyVersion as it was when we are adding API for core-only. So in this case, I guess that we have to revert this change, and that might solve your build issue.

File: src/System.Collections.Concurrent/pkg/System.Collections.Concurrent.pkgproj [hunk 13]
```diff
@@ -2,8 +2,8 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
-    <ProjectReference Include="..\ref\System.Collections.Concurrent.csproj">
-      <SupportedFramework>net46;netcore50;netcoreapp1.0;$(AllXamarinFrameworks)</SupportedFramework>
+    <ProjectReference Include="..\ref\System.Collections.Concurrent.builds">
+      <SupportedFramework>netcoreapp1.1;net463;$(AllXamarinFrameworks)</SupportedFramework>
```
Comment (joperezr) [inline src/System.Collections.Concurrent/pkg/System.Collections.Concurrent.pkgproj]: >net463 [](start = 40, length = 6)

Let's remove `net463` from here since we are not exposing any new API for it.

Comment (weshaggard) [inline src/System.Collections.Concurrent/pkg/System.Collections.Concurrent.pkgproj]: I think this should also fix your current build issue.

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj [hunk 14]
```diff
@@ -2,9 +2,10 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.10.0</AssemblyVersion>
+    <AssemblyVersion Condition="'$(TargetGroup)'=='netstandard1.3'">4.0.10.0</AssemblyVersion>
```
Comment (joperezr) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: As I said on my previous comment on the dir.props file, I misguided you to update this AssemblyVersion, since it looks like we only want to update the AssemblyVersion whenever we expose new API for `netstandard` and not when we only expose it for `netcoreapp`. That means that this should go back to being: `<AssemblyVersion>4.0.10.0</AssemblyVersion>` That should fix your build issue.

File: src/System.Collections.Concurrent/ref/project.json (hunk 8 above)
Comment (joperezr) [inline src/System.Collections.Concurrent/ref/project.json]: no, there is no new API for netstandard so this is fine.

File: src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj (hunk 9 above)
Comment (joperezr) [inline src/System.Collections.Concurrent/ref/System.Collections.Concurrent.csproj]: yes, if we x-compile for netstandard1.3

File: src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryExtensions.cs (hunk 10 above)
Comment (justinvp) [inline src/System.Collections.Concurrent/tests/ConcurrentDicionary/ConcurrentDictionaryExtensions.cs]: It wasn't as clean to separate out the tests into a separate file in this case. Steve suggested the extension methods. https://github.com/dotnet/corefx/pull/13828#discussion_r88785528

File: src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds (hunk 11 above)
Comment (justinvp) [inline src/System.Collections.Concurrent/tests/System.Collections.Concurrent.Tests.builds]: I had that originally, but removed based on feedback from Jose as it's the default. https://github.com/dotnet/corefx/pull/13828#discussion_r88946174

File: src/System.Collections.Concurrent/dir.props (hunk 12 above)
Comment (joperezr) [inline src/System.Collections.Concurrent/dir.props]: Yes, I believe so. cc: @ericstj @stephentoub Should we log an issue to track reverting that change and bring the assembly version down again given that api changes were only for netcoreapp1.1?

**Comment by justinvp** [comment]
I reverted the version bump per @joperezr's feedback (https://github.com/dotnet/corefx/pull/13828#discussion_r90931863). CI all green now.

**Comment by justinvp** [comment]
Unless there's any other feedback, this should be good to merge.
